core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala

```diff
@@ -70,6 +86,258 @@ private[spark] object ResourceUtils extends Logging {
   // internally we currently only support addresses, so its just an integer count
   val AMOUNT = "amount"
 
+  /**
+   * Assign resources to workers from the same host to avoid address conflict.
+   * @param conf SparkConf
+   * @param componentName spark.driver / spark.worker
+   * @param resources the resources found by worker/driver on the host
+   * @param resourceRequirements the resource requirements asked by the worker/driver
+   * @param pid the process id of worker/driver to acquire resources.
+   * @return allocated resources for the worker/driver or throws exception if can't
+   *         meet worker/driver's requirement
+   */
+  def acquireResources(
+      conf: SparkConf,
+      componentName: String,
+      resources: Map[String, ResourceInformation],
+      resourceRequirements: Seq[ResourceRequirement],
+      pid: Int)
+    : Map[String, ResourceInformation] = {
+    if (resourceRequirements.isEmpty) {
```

Comment:
this function is hard to read might be nice to break it up into smaller chunks

Reply:
Have added more comments to make it more readable, but sorry have't break it up. I tried it, but didn't find a good way. Let me think it a bit more.